File: src/App/Core/FileEncryptor.cs
using System;
using System.IO;
using System.Security.Cryptography;

namespace SteamDesktopAuthenticator.Core
{
    /// <summary>
    /// Ported 1:1 from the original SDA FileEncryptor.cs. Behavior and on-disk format are
    /// unchanged so existing .maFile / manifest.json data remains fully compatible:
    ///   - PBKDF2 (RFC2898), 50,000 iterations, 32-byte derived key
    ///   - AES-256, CBC mode, PKCS7 padding
    ///   - 8-byte random salt, 16-byte random IV, both base64-encoded and stored per-entry
    ///
    /// The only change from the original is using System.Security.Cryptography.Aes.Create()
    /// instead of the obsolete RijndaelManaged class, which is not available outside .NET
    /// Framework/Windows. With IV length 16 and key length 32, Rijndael and AES are byte-for-byte
    /// identical algorithms, so this does not change the format or break existing files.
    /// </summary>
    public static class FileEncryptor
    {
        private const int PBKDF2_ITERATIONS = 50000;
        private const int SALT_LENGTH = 8;
        private const int KEY_SIZE_BYTES = 32;
        private const int IV_LENGTH = 16;

        public static string GetRandomSalt()
        {
            byte[] salt = new byte[SALT_LENGTH];
            RandomNumberGenerator.Fill(salt);
            return Convert.ToBase64String(salt);
        }

        public static string GetInitializationVector()
        {
            byte[] iv = new byte[IV_LENGTH];
            RandomNumberGenerator.Fill(iv);
            return Convert.ToBase64String(iv);
        }

        private static byte[] GetEncryptionKey(string password, string salt)
        {
            if (string.IsNullOrEmpty(password)) throw new ArgumentException("Password is empty");
            if (string.IsNullOrEmpty(salt)) throw new ArgumentException("Salt is empty");

            using var pbkdf2 = new Rfc2898DeriveBytes(password, Convert.FromBase64String(salt), PBKDF2_ITERATIONS, HashAlgorithmName.SHA1);
            return pbkdf2.GetBytes(KEY_SIZE_BYTES);
        }

        public static string? DecryptData(string password, string passwordSalt, string iv, string encryptedData)
        {
            if (string.IsNullOrEmpty(password)) throw new ArgumentException("Password is empty");
            if (string.IsNullOrEmpty(passwordSalt)) throw new ArgumentException("Salt is empty");
            if (string.IsNullOrEmpty(iv)) throw new ArgumentException("Initialization Vector is empty");
            if (string.IsNullOrEmpty(encryptedData)) throw new ArgumentException("Encrypted data is empty");

            byte[] cipherText = Convert.FromBase64String(encryptedData);
            byte[] key = GetEncryptionKey(password, passwordSalt);

            using var aes = Aes.Create();
            aes.IV = Convert.FromBase64String(iv);
            aes.Key = key;
            aes.Padding = PaddingMode.PKCS7;
            aes.Mode = CipherMode.CBC;

            try
            {
                using var decryptor = aes.CreateDecryptor(aes.Key, aes.IV);
                using var msDecrypt = new MemoryStream(cipherText);
                using var csDecrypt = new CryptoStream(msDecrypt, decryptor, CryptoStreamMode.Read);
                using var srDecrypt = new StreamReader(csDecrypt);
                return srDecrypt.ReadToEnd();
            }
            catch (CryptographicException)
            {
                return null;
            }
        }

        public static string EncryptData(string password, string passwordSalt, string iv, string plaintext)
        {
            if (string.IsNullOrEmpty(password)) throw new ArgumentException("Password is empty");
            if (string.IsNullOrEmpty(passwordSalt)) throw new ArgumentException("Salt is empty");
            if (string.IsNullOrEmpty(iv)) throw new ArgumentException("Initialization Vector is empty");
            if (string.IsNullOrEmpty(plaintext)) throw new ArgumentException("Plaintext data is empty");

            byte[] key = GetEncryptionKey(password, passwordSalt);

            using var aes = Aes.Create();
            aes.Key = key;
            aes.IV = Convert.FromBase64String(iv);
            aes.Padding = PaddingMode.PKCS7;
            aes.Mode = CipherMode.CBC;

            using var encryptor = aes.CreateEncryptor(aes.Key, aes.IV);
            using var msEncrypt = new MemoryStream();
            using (var csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write))
            using (var swEncrypt = new StreamWriter(csEncrypt))
            {
                swEncrypt.Write(plaintext);
            }
            return Convert.ToBase64String(msEncrypt.ToArray());
        }
    }
}

File: Application/app/src/main/java/com/sda/mobile/crypto/FileEncryptor.kt
package com.sda.mobile.crypto

import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Port of the desktop app's Core/FileEncryptor.cs. Same on-disk format, so a .maFile encrypted
 * by this app decrypts correctly in the desktop app and vice versa:
 *   - PBKDF2 (RFC2898/PBKDF2WithHmacSHA1), 50,000 iterations, 32-byte derived key
 *   - AES-256, CBC mode, PKCS7 (PKCS5, byte-identical for a 16-byte block) padding
 *   - 8-byte random salt, 16-byte random IV, both base64-encoded and stored per-entry
 */
object FileEncryptor {
    private const val PBKDF2_ITERATIONS = 50000
    private const val SALT_LENGTH_BYTES = 8
    private const val KEY_SIZE_BITS = 256
    private const val IV_LENGTH_BYTES = 16

    private val secureRandom = SecureRandom()

    fun getRandomSalt(): String {
        val salt = ByteArray(SALT_LENGTH_BYTES)
        secureRandom.nextBytes(salt)
        return Base64.encodeToString(salt, Base64.NO_WRAP)
    }

    fun getInitializationVector(): String {
        val iv = ByteArray(IV_LENGTH_BYTES)
        secureRandom.nextBytes(iv)
        return Base64.encodeToString(iv, Base64.NO_WRAP)
    }

    private fun getEncryptionKey(password: String, saltBase64: String): ByteArray {
        require(password.isNotEmpty()) { "Password is empty" }
        require(saltBase64.isNotEmpty()) { "Salt is empty" }

        val salt = Base64.decode(saltBase64, Base64.DEFAULT)
        val spec = PBEKeySpec(password.toCharArray(), salt, PBKDF2_ITERATIONS, KEY_SIZE_BITS)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1")
        return factory.generateSecret(spec).encoded
    }

    /** Returns null on a bad passkey (padding/MAC failure), matching the desktop behavior. */
    fun decryptData(password: String, passwordSalt: String, iv: String, encryptedData: String): String? {
        require(password.isNotEmpty()) { "Password is empty" }
        require(passwordSalt.isNotEmpty()) { "Salt is empty" }
        require(iv.isNotEmpty()) { "Initialization Vector is empty" }
        require(encryptedData.isNotEmpty()) { "Encrypted data is empty" }

        val cipherText = Base64.decode(encryptedData, Base64.DEFAULT)
        val key = getEncryptionKey(password, passwordSalt)

        return try {
            val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            cipher.init(
                Cipher.DECRYPT_MODE,
                SecretKeySpec(key, "AES"),
                IvParameterSpec(Base64.decode(iv, Base64.DEFAULT))
            )
            String(cipher.doFinal(cipherText), Charsets.UTF_8)
        } catch (e: Exception) {
            null
        }
    }

    fun encryptData(password: String, passwordSalt: String, iv: String, plaintext: String): String {
        require(password.isNotEmpty()) { "Password is empty" }
        require(passwordSalt.isNotEmpty()) { "Salt is empty" }
        require(iv.isNotEmpty()) { "Initialization Vector is empty" }
        require(plaintext.isNotEmpty()) { "Plaintext data is empty" }

        val key = getEncryptionKey(password, passwordSalt)
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(
            Cipher.ENCRYPT_MODE,
            SecretKeySpec(key, "AES"),
            IvParameterSpec(Base64.decode(iv, Base64.DEFAULT))
        )
        return Base64.encodeToString(cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8)), Base64.NO_WRAP)
    }
}
